src/modules/flow/si114x/si114x.c

```diff
@@ -0,0 +1,333 @@
+
+    REG_PS_LED21 = 0x0f,
+    REG_PS_LED3 = 0x10,
+
+    REG_UCOEF0 = 0x13,
+    REG_UCOEF1 = 0x14,
+    REG_UCOEF2 = 0x15,
+    REG_UCOEF3 = 0x16,
+    REG_PARAM_WR = 0x17,
+    REG_COMMAND = 0x18,
+
+    REG_RESPONSE = 0x20,
+    REG_IRQ_STATUS = 0x21,
+    REG_ALS_VIS_DATA0 = 0x22,
+    REG_ALS_VIS_DATA1 = 0x23,
+    REG_ALS_IR_DATA0 = 0x24,
+    REG_ALS_IR_DATA1 = 0x25,
+    REG_PS1_DATA0 = 0x26,
+    REG_PS1_DATA1 = 0x27,
+    REG_PS2_DATA0 = 0x28,
+    REG_PS2_DATA1 = 0x29,
+    REG_PS3_DATA0 = 0x2a,
+    REG_PS3_DATA1 = 0x2b,
+    REG_AUX_UVINDEX0 = 0x2c,
+    REG_AUX_UVINDEX1 = 0x2d,
+    REG_PARAM_READ = 0x2e,
+
+    REG_CHIP_STAT = 0x30,
+
+    REG_ANA_IN_KEY0 = 0x3b,
+    REG_ANA_IN_KEY1 = 0x3c,
+    REG_ANA_IN_KEY2 = 0x3d,
+    REG_ANA_IN_KEY3 = 0x3e
+};
+
+/*
+ * Parameter memory (PARAM)
+ */
+enum SI114X_PARAM {
+    PARAM_I2C_ADDDR = 0x00,
+    PARAM_CHLIST = 0x01,
+    PARAM_PSLED12_SEL = 0x02,
+    PARAM_PSLED3_SEL = 0x03,
+
+    PARAM_PS_ENCODING = 0x05,
+    PARAM_ALS_ENCODING = 0x06,
+    PARAM_PS1_ADCMUX = 0x07,
+    PARAM_PS2_ADCMUX = 0x08,
+    PARAM_PS3_ADCMUX = 0x09,
+    PARAM_PS_ADC_COUNT = 0x0a,
+    PARAM_PS_ADC_GAIN = 0x0b,
+    PARAM_PS_ADC_MISC = 0x0c,
+
+    PARAM_ALS_IR_ADCMUX = 0x0e,
+    PARAM_AUX_ADCMUX = 0x0f,
+    PARAM_ALS_VIS_ADC_COUNT = 0x10,
+    PARAM_ALS_VIS_ADC_GAIN = 0x11,
+    PARAM_ALS_VIS_ADC_MISC = 0x12,
+
+    PARAM_LED_REC = 0x1c,
+    PARAM_ALS_IR_ADC_COUNT = 0x1d,
+    PARAM_ALS_IR_ADX_GAIN = 0x1e,
+    PARAM_ALS_IR_ADC_MISC = 0x1f
+};
+
+/*
+ * Commands (written to the REG_COMMAND register)
+ */
+enum SI114X_CMD {
+    CMD_NOOP = 0x00, // clear RESPONSE reg
+    CMD_RESET = 0x01,
+    CMD_BUSADDR = 0x02,
+
+    CMD_PS_FORCE = 0x05,
+    CMD_GET_CAL = 0x12,
+    CMD_ALS_FORCE = 0x06,
+    CMD_PSALS_FORCE = 0x07,
+
+    CMD_PS_PAUSE = 0x09,
+    CMD_ALS_PAUSE = 0x0a,
+    CMD_PSALS_PAUSE = 0x0b,
+
+    CMD_PS_AUTO = 0x0d,
+    CMD_ALS_AUTO = 0x0e,
+    CMD_PSALS_AUTO = 0x0f,
+
+    CMD_PARAM_QUERY = 0x80, // or'd with PARAM_T value
+    CMD_PARAM_SET = 0xa0  // or'd with PARAM_T value
+};
+
+/*
+ * Channel List enable bits
+ */
+enum SI114X_CHLIST_BITS {
+    CHLIST_EN_PS1 = 0x01, // proximity sense 1-3
+    CHLIST_EN_PS2 = 0x02,
+    CHLIST_EN_PS3 = 0x04,
+
+    CHLIST_EN_ALS_VIS = 0x10, // ambient light sense
+    CHLIST_EN_ALS_IR = 0x20,
+    CHLIST_EN_AUX = 0x40, // AUX sense
+    CHLIST_EN_UV = 0x80  // UV sense
+};
+
+struct si114x_data {
+    struct sol_i2c *context;
+    struct sol_timeout *timer;
+    bool fully_initialized;
+    unsigned int init_step;
+    int pendent_calls;
+    uint16_t read_data;
+    struct sol_flow_node *node;
+};
+
+#define SI114X_I2C_BUS 0
+#define SI114X_DEFAULT_I2C_ADDR 0x60
+#define SI114X_HW_KEY 0x17
+
+struct i2c_initialization_data {
+    uint8_t reg;
+    uint8_t value;
+    const char *error_str;
+} initialization_data[] = {
+    { .reg = REG_MEAS_RATE0, .value = 0, .error_str = "Couldn't reset the REG_MEAS_RATE0 register" },
```

Comment:
Why `error_str` isn't just `REG_MEAS_RATE0`, and the function that process the initialization data vector builds an error message such as `Could not write %value to register %register`?